controllers/helmrelease_controller_release.go

```diff
@@ -0,0 +1,426 @@
+/*
+Copyright 2020 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/fluxcd/pkg/apis/meta"
+	"github.com/fluxcd/pkg/runtime/events"
+	"github.com/go-logr/logr"
+	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/strvals"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	v2 "github.com/fluxcd/helm-controller/api/v2beta1"
+	"github.com/fluxcd/helm-controller/internal/runner"
+	"github.com/fluxcd/helm-controller/internal/util"
+)
+
+func (r *HelmReleaseReconciler) reconcileRelease(ctx context.Context, log logr.Logger, run *runner.Runner, hr *v2.HelmRelease) (ctrl.Result, error) {
+	// Given we are in control over creating the chart, we should
+	// return an error as we either need to reconcile again to recreate
+	// it, or reattempt because we encountered a transient error.
+	hc, err := r.getHelmChart(ctx, hr)
+	if err != nil {
+		err = fmt.Errorf("failed to get HelmChart for resource: %w", err)
+		v2.HelmReleaseNotReady(hr, v2.ArtifactFailedReason, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Check if the chart has an artifact. If there is no artifact,
+	// record the fact that we observed this but do not return an error.
+	// Our watcher should observe the chart becoming ready, and queue
+	// a new reconciliation.
+	if hc.GetArtifact() == nil {
+		msg := fmt.Sprintf("HelmChart '%s/%s' has no artifact", hc.GetNamespace(), hc.GetName())
+		v2.HelmReleaseNotReady(hr, v2.ArtifactNotReadyReason, msg)
+		return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+	}
+
+	// Compose the values. As race conditions may happen due to
+	// unordered applies (due to e.g. a HelmRelease with a reference
+	// to a Secret is applied before the Secret itself), we expect
+	// most errors to be transient and return them.
+	values, err := r.composeValues(ctx, log, hr)
+	if err != nil {
+		v2.HelmReleaseNotReady(hr, v2.InitFailedReason, err.Error())
+		r.event(hr, hr.Status.LastAttemptedRevision, events.EventSeverityError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Observe the last release. If this fails, we likely encountered a
+	// transient error and should return it to requeue a reconciliation.
+	rls, err := run.ObserveLastRelease(hr)
```

Comment:
`ObserveLastRelease` retrieves the last release from the Helm storage, and this may return a manually made release, the "cached" version is returned by `GetLastPersistedRelease`.

Reply:
Ah, I got confused between `ObserveLastRelease` and `GetLastObservedRelease` whose names make it seem like they are more closely related than they are.